pkg/sql/catalog/descs/uncommitted_descriptors.go

```diff
@@ -111,17 +110,19 @@ func (ud *uncommittedDescriptors) getUncommittedByName(
 	return nil
 }
 
-// iterateNewVersionByID applies fn to each lease.IDVersion from the original
-// descriptor, if it exists, for each uncommitted descriptor, in ascending
-// sequence of IDs.
-func (ud *uncommittedDescriptors) iterateNewVersionByID(
-	fn func(originalVersion lease.IDVersion) error,
+// iterateOriginalByID applies fn to each original descriptor for which an
+// uncommitted non-new descriptor exists (new descriptors don't have an original
+// descriptor).
+func (ud *uncommittedDescriptors) iterateOriginalByID(
+	fn func(original catalog.Descriptor) error,
 ) error {
 	return ud.uncommitted.IterateByID(func(entry catalog.NameEntry) error {
-		if o := ud.original.Get(entry.GetID()); o != nil {
-			return fn(lease.NewIDVersionPrev(o.GetName(), o.GetID(), o.(catalog.Descriptor).GetVersion()))
```

Comment:
The bug proper was that `fn` was being called a bit too liberally here. The regression was introduced in this PR, I highlighted the line with the relevant part of the old logic: https://github.com/cockroachdb/cockroach/pull/87067/files#diff-74473a29d93e28d8f048009aed817c687c38ab748f111c95fba1ec148f7c0768L664

I fixed this and pushed a more verbose version of this code down to the call site in the hopes of clarifying what it's supposed to do.